### DISCUSSION TITLE
Pair Repulsion in MACE

### DISCUSSION THREAD

**Discussion by nicdieugenio** [ACEsuit/mace#992] [Q&A] [answered]
Dear all,
A follow-up on Discussion #982:
I am testing how the pair_repulsion + distance_transform flag works in MACE. I have noticed that I was getting some unphysical results when training the potential with different settings of interaction block and interaction_first block options and then computing pair repulsion considering a diatomic molecule with:

model_path = "MACE_model_stagetwo.model"
calc = MACECalculator(model_paths=[model_path], device="cuda")
elements = ['Cu']
for i,el1 in enumerate(elements):
    for j,el2 in enumerate(elements[i:],start=i):
        print(el1,el2)
        foutname = "e_vs_dist_MACE_%s-%s" % (el1,el2)
        fff = open(foutname, 'w')
        for m in range(0,1001,1):
          mol = "%s%s" % (el1,el2)
          d = 0.1+m/100
          atoms = Atoms(mol, positions=[(0,0,0),(0,0,d)])
          atoms.set_calculator(calc)
          en = atoms.get_potential_energy()
          fff.write("%.2f %.10f\n" % (d,en))
        fff.close() 

I thus did the following:

Starting from a dataset of Cu structures (with volume, strains and random displacements scaling) I computed a different potential for each combination of:

- distance_transform: Agnesi, Soft
- interaction blocks: RealAgnosticResidualInteractionBlock, RealAgnosticAttResidualInteractionBlock, RealAgnosticInteractionBlock, RealAgnosticDensityInteractionBlock, RealAgnosticDensityResidualInteractionBlock
- interaction first blocks: RealAgnosticResidualInteractionBlock, RealAgnosticInteractionBlock, RealAgnosticDensityInteractionBlock, RealAgnosticDensityResidualInteractionBlock

keeping all the other input parameters fixed, as in:

    --name="MACE_${DISTANCE_TRANSFORM}_${INTER}_${INTER_FIRST}" \\
    --train_file="Cu-reduced.xyz" \\
    --valid_fraction=0.05 \\
    --E0s="{29: -1306.07919131771}" \\
    --model="MACE" \\
    --hidden_irreps="128x0e" \\
    --num_interactions=2 \\
    --max_L=0 \\
    --r_max=6.5 \\
    --batch_size=6 \\
    --max_num_epochs=250 \\
    --ema \\
    --ema_decay=0.99 \\
    --amsgrad \\
    --default_dtype="float32" \\
    --pair_repulsion \\
    --distance_transform="${DISTANCE_TRANSFORM}" \\
    --interaction ${INTER} \\
    --interaction_first ${INTER_FIRST} \\
    --device=cuda \\
    --enable_cueq=True \\
    --seed=12 \\
    --swa \\
    --start_swa=200 \\
    --energy_key="REF_energy" \\
    --forces_key="REF_forces" \\
    --stress_key="Stress" \\
    --scaling="rms_forces_scaling" \\
    --log_level="INFO" \\
    --loss="weighted" \\
    --forces_weight=10.0 \\
    --energy_weight=1.0 \\
    --clip_grad=5.0 \\
    --weight_decay=5e-7 \\
    --restart_latest
    
I obtained a total of 40 different potentials and, among those that did not fail for compatibility issues, I computed Cu-Cu pair repulsion and compared with ZBL and all electron DFT calculations (DMOL) from https://doi.org/10.1103/PhysRevA.111.032818 (bold lines).

Since I noticed that when using "Soft" the behavior of pair repulsion curves was showing several minima and therefore was not physical, I restricted to only Agnesi distance_transform and picked the most promising combinations that are shown in: 

![pair_rep](https://github.com/user-attachments/assets/e7bafa6d-1643-4f6f-9303-d42079b4ebc0)

Why is there a difference in behavior according to these options? Are the interaction and interaction first options also connected to how the distance_transform/ZBL works? Are there any references I can look at to understand the best setup?

In addition, if I enlarge the y range, I notice that the potentials that I trained never join the ZBL (green bold line) at very small distances:

![pair_plots](https://github.com/user-attachments/assets/db9b258c-7514-4848-861b-588370a5dd13)

For completeness, the covalent radius of Cu is 1.32 Å (as computed from the ase.data.covalent_radii function used in radial.py) and the smallest pair distance in my dataset is 1.07 Å.
Why at short distances my potentials do not coincide with the ZBL? Am I missing something? 

Best,

Niccolò Di Eugenio

**Comment by ilyes319** [comment, marked as answer]
Hello @nicdieugenio,

Different block will end up learning different radial basis, hence why the look different. 
To choose which block to use,  I would recommend a few options:
- The usual setup is to use the first interaction: RealAgnosticInteractionBlock and the second: RealAgnosticResidualInteractionBlock. This will enable your model to predict the true isolated atoms energy. If you are not using true E0s but average, you can use RealAgnosticResidualInteractionBlock for the two layers.
- The new setup we use for foundation models contains a new trick called density normalization, that enables better extrapolation to high temperature and pressures. This will described in our updates pre print to the MP-0 paper. We use for the first layer: RealAgnosticDensityInteractionBlock and for the second layer: RealAgnosticDensityResidualInteractionBlock. You can give a try, but be aware that we don't have much experience of it on the low data regime, it was made for big fits mainly.

The ZBL we have in the code is a slightly different than the lammps ZBL, if you used that to do your plot hence why they don't join. I have change the parameters slighty to enable smoother training.

**Comment by nicdieugenio** [comment]
Dear @ilyes319,
Thank you very much for the explanation and for the advice, I will give it a try!
Best,

Niccolò
